libraries/azure_backend.rb

```diff
@@ -0,0 +1,788 @@
+require 'rubygems'
+require 'bundler/setup'
+require 'pp'
+require_relative 'backend/azure_connection'
+require 'rspec/expectations'
+require 'facets/string'
+require 'dotenv'
+
+# `../.env` is the correct relative path from this file, however,
+#   InSpec runs this file from the root directory, hence the `.env`.
+Dotenv.load('.env', '../.env')
+ENV_HASH = ENV.map { |k, v| [k.downcase, v] }.to_h
+
+# Base class for Azure resources.
+#
+# Provides:
+#  - Connection to Azure Rest API.
+#  - Parameter validation.
+#  - Short description of resources. This includes resource_id.
+#  - Detailed (long) description of resources upon providing a resource_id.
+#  - Latest or default api version for a resource provider endpoint.
+#  - Rescuing invalid api version error by using the suggested api version in the error message.
+#  - Creating resource methods dynamically.
+#
+class AzureResourceBase < Inspec.resource(1)
+  def initialize(opts = {})
+    raise ArgumentError, 'Parameters must be provided in an Hash object.' unless opts.is_a?(Hash)
+    @opts = opts
+
+    # Populate client_args to specify AzureConnection
+    #
+    # The valid client args (all of them are optional):
+    #  - endpoint: [String] azure_cloud (default), azure_china_cloud, azure_us_government, azure_german_cloud
+    #  - azure_retry_limit: [Integer] Maximum number of retries (default - 2)
+    #  - azure_retry_backoff: [Integer] Pause in seconds between retries (default - 0)
+    #  - azure_retry_backoff_factor: [Integer] The amount to multiply each successive retry's interval amount
+    #     by in order to provide back-off (default - 1)
+    @client_args = {}
+    # If not provided, the endpoint will be the Global Cloud portal.
+    # https://azure.microsoft.com/en-gb/global-infrastructure/
+    @client_args[:endpoint] = @opts[:endpoint] || ENV_HASH['endpoint'] || 'azure_cloud'
+    unless AzureEnvironments::ENDPOINTS.key?(@client_args[:endpoint])
+      raise ArgumentError, "Invalid endpoint: `#{@client_args[:endpoint]}`."\
+        " Expected one of the following options: #{AzureEnvironments::ENDPOINTS.keys}."
+    end
```

Comment:
It doesn't seem like it would be possible to ever hit this ArgumentError, considering it always will default to `azure_cloud` when no opt or ENV var is provided

Reply:
A wrong endpoint can be provided by the user:
`azure_virtual_machines(endpoint: 'not_existing')` will raise this error.